frame/contracts/src/exec.rs

```diff
@@ -1016,6 +1026,19 @@ where
 		)
 	}
 
+	fn set_code_hash(&mut self, hash: CodeHash<Self::T>) -> Result<(), DispatchError> {
+		let top_frame = &mut self.top_frame_mut();
+		let prev_hash = top_frame.contract_info().code_hash.clone();
+		if !Self::code_exists(&hash) {
+			return Err(DispatchError::CannotLookup)
+		}
```

Comment:
Don't do this check explicitly. Do the changes to `increment_refcount` I pointed out and then error out when it fails. Also, move those refcount operations up so we can error out earlier.